src/concrete/ml/sklearn/base.py

```diff
@@ -1690,3 +1694,382 @@ def predict_proba(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) ->
+                enc_arr = fhe_array(arr)
+                return enc_arr
+
+            def scatter1d(x, v, indices):
+                """Rearrange elements of `x` with values from `v` at the specified `indices`.
+
+                Args:
+                    x (numpy.ndarray): The encrypted input array in which items will be updated
+                    v (numpy.ndarray): The array containing values to be inserted into `x`
+                        at the specified `indices`.
+                    indices (numpy.ndarray): The indices indicating where to insert the elements
+                        from `v` into `x`.
+
+                Returns:
+                    numpy.ndarray: The updated encrypted `x`
+                """
+                for idx, i in enumerate(indices):
+                    x[i] = v[idx]
+                return x
+
+            comparisons = numpy.zeros(x.shape)
+            labels = labels + fhe_zeros(labels.shape)
+
+            n, k = x.size, self.n_neighbors
+            ln2n = int(numpy.ceil(numpy.log2(n)))
+
+            # Number of stages
+            for t in range(ln2n - 1, -1, -1):
+                p = 2**t
+                r = 0
+                # d: Length of the bitonic sequence
+                d = p
+
+                for bq in range(ln2n - 1, t - 1, -1):
+                    q = 2**bq
+                    # Determine the range of indexes to be compared
+                    range_i = numpy.array(
+                        [i for i in range(0, n - d) if i & p == r and comparisons[i] < k]
+                    )
+                    if len(range_i) == 0:
+                        # Edge case, for k=1
+                        continue
+
+                    # Select 2 bitonic sequences `a` and `b` of length `d`
+                    # a = x[range_i]: first bitonic sequence
+                    # a_i = idx[range_i]: Indexes of a_i elements in the original x
+                    a = gather1d(x, range_i)
+                    # a_i = gather1d(idx, range_i)
+                    # b = x[range_i + d]: Second bitonic sequence
+                    # b_i = idx[range_i + d]: Indexes of b_i elements in the original x
+                    b = gather1d(x, range_i + d)
+                    # b_i = gather1d(idx, range_i + d)
+
+                    labels_a = gather1d(labels, range_i)  #
+                    labels_b = gather1d(labels, range_i + d)  # idx[range_i + d]
+
+                    # Select max(a, b)
+                    diff = a - b
+                    max_x = a + numpy.maximum(0, b - a)
+
+                    # Swap if a > b
+                    # x[range_i] = max_x(a, b): First bitonic sequence gets min(a, b)
+                    x = scatter1d(x, a + b - max_x, range_i)
+                    # x[range_i + d] = min(a, b): Second bitonic sequence gets max(a, b)
+                    x = scatter1d(x, max_x, range_i + d)
+
+                    # Max index selection
+                    sign = diff <= 0
```

Comment:
This is not the sign but a boolean I guess. Why isn't this done right after computing diff?

Have you tried the CP comparison optimization by replacing
```python
diff = a - b
sign = diff <= 0
```
to
```python
sign = a <= b
```
?

Reply:
It's indeed a boolean that tells us if a is greater than b.
I'll change the naming.

Reply:
Maybe you can just check on a simple example if you have any time improvement. Otherwise you can leave it.

Reply:
if you don't have the time right now worth an issue then I think because this might be a good thing to try anyway

Reply:
I'll do it in a separate PR.
Some testings are needed.

Reply:
Great can you create an issue for this?